### PR TITLE
Gym Usage: Line Colors

### PR DESCRIPTION
**Pull request by BlobbyBob**
:sparkles: Lines have an optional color associated with them
:sparkles: Line Form allows setting the color if in gym mode and suggests 10 (currently hardcoded) colors
:sparkles: UI respects the line colors whereever set and falls back to the instance setting otherwise

### CODE REVIEW

File: client/src/app/modules/line/line-form/line-form.component.html [hunk 1]
```diff
@@ -70,6 +70,35 @@
             </p-editor>
           </div>
 
+          <ng-container *isGymMode="true">
+            <div class="flex flex-column gap-2" lcControlGroup>
```
Comment (dorthrithil) [inline client/src/app/modules/line/line-form/line-form.component.html:74]: I would suggest putting this in a separate component that implements the `ControlValueAccessor`. This way we can keep the color logic out of the form and make the color selector reusable (maybe later we want to choose colors for rectangles in maps or similar)

File: client/src/app/modules/line/line-form/line-form.component.ts [hunk 2]
```diff
@@ -397,4 +405,6 @@ export class LineFormComponent implements OnInit {
   public deleteLineVideoControl(index: number) {
     (this.lineForm.get('videos') as FormArray).removeAt(index);
   }
+
+  protected readonly console = console;
```
Comment (dorthrithil) [inline client/src/app/modules/line/line-form/line-form.component.ts:409]: Remove debug var

File: client/src/app/modules/shared/components/color-square/color-square.component.html [hunk 3]
```diff
@@ -0,0 +1,3 @@
+<ng-container>
```
Comment (dorthrithil) [inline client/src/app/modules/shared/components/color-square/color-square.component.html:1]: container not needed

File: client/src/app/utility/misc/color.ts [hunk 4]
```diff
@@ -13,3 +13,36 @@ export function getRgbObject(raw: string): any {
...
+    r: parseInt(color.slice(1, 3), 16),
+    g: parseInt(color.slice(3, 5), 16),
+    b: parseInt(color.slice(5, 7), 16),
+    brightness: 0,
+  };
+  ret.brightness = (ret.r * 299 + ret.g * 587 + ret.b * 114) / 1000;
+  return ret;
+}
+
+export function highlightColor(color?: string): string | undefined {
```
Comment (dorthrithil) [inline client/src/app/utility/misc/color.ts:28]: For the functions added here, some docstring would be good

**Review by dorthrithil — CHANGES_REQUESTED** [review]
- No color tests
- I think there is no option to set a color back to null (this is, making it take the default color, even if you change that)
- The archive controls should be gym mode only. Could be validated on backend too, but I think hiding it is enough as we are not exposing any sensitive api to non-moderators

File: client/src/app/modules/line/line-form/line-form.component.ts (hunk 2 above)
Comment (BlobbyBob) [reply client/src/app/modules/line/line-form/line-form.component.ts:409]: I removed all console calls, but how thought this sneaks into the code.
I hate it when my IDE is over eager and adds such useless stuff ...

**Comment by BlobbyBob** [comment]
It's almost finished, but I currently have an error which I did not find a fix yet:
```
NullInjectorError: R3InjectorError(CoreModule)[NgControl -> NgControl]: 
  NullInjectorError: No provider for NgControl!
```

It occurs in `advanced-color-picker.component.ts:65`. However, I don't see any difference to the other components in `shared/forms/controls` where it works.

I think you are a bit more Angular experienced and might be able to find the issue faster than me?

**Comment by dorthrithil** [comment]
> It's almost finished, but I currently have an error which I did not find a fix yet:
> 
> ```
> NullInjectorError: R3InjectorError(CoreModule)[NgControl -> NgControl]: 
>   NullInjectorError: No provider for NgControl!
> ```
> 
> It occurs in `advanced-color-picker.component.ts:65`. However, I don't see any difference to the other components in `shared/forms/controls` where it works.
> 
> I think you are a bit more Angular experienced and might be able to find the issue faster than me?

Just from a quick look I think this is because of how it's used: `<lc-advanced-color-picker/>`.

Try adding the `formControlName` directive. This way [`NgControl` is provided ](https://github.com/angular/angular/blob/7d0ba0cac85220cbbe4044667a51e5b95512f5d6/packages/forms/src/directives/reactive_directives/form_control_name.ts#L46) and the control knows which control in the form it belongs to.

```
<lc-advanced-color-picker formControlName="color"/>
```

If this isn't the issue, I can have a deeper look.

**Comment by BlobbyBob** [comment]
> Just from a quick look I think this is because of how it's used: `<lc-advanced-color-picker/>`.

Ah yeah thanks, that fixed it. I was too focused that the issue is with the component itself.

**Review by dorthrithil — APPROVED** [review]

**Pull request merged**